File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/session/SecureToken.java
package solutions.digamma.damas.jcr.session;

import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.inspection.NotNull;

import java.math.BigInteger;
import java.security.SecureRandom;

/**
 * Statistically secure token.
 *
 * @author Ahmad Shahwan
 */
public class SecureToken implements Token {

    @NotNull
    private final String token;

    private static final SecureRandom RANDOM = new SecureRandom();

    /**
     * Constructor.
     */
    public SecureToken() {
        this.token = SecureToken.nextToken();
    }

    @NotNull
    public String getSecret() {
        return this.token;
    }

    private static String nextToken() {
        return new BigInteger(132, RANDOM).toString(64);
    }

    @Override
    public boolean equals(Object other) {
        /* Two token with no secret do not equal each other. */
        return
            other instanceof Token &&
            this.getSecret() != null &&
            this.getSecret().equals(((Token) other).getSecret());
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/session/SecureToken.kt
package solutions.digamma.damas.jcr.session

import solutions.digamma.damas.session.UserToken

import java.math.BigInteger
import java.security.SecureRandom
import java.util.Date

/**
 * Statistically secure token.
 *
 * @author Ahmad Shahwan
 */
internal class SecureToken(
    private val login: String,
) : UserToken {

    private val token: String
    private val creation: Date

    /**
     * Constructor.
     */
    init {
        this.token = SecureToken.nextToken()
        this.creation = Date()
    }

    override fun getSecret() = this.token

    /**
     * Two token with no secret do not equal each other.
     *
     * @param other object to compare with
     * @return equality
     */
    override fun equals(other: Any?): Boolean =
        other is UserToken &&
                this.secret == other.secret &&
                this.userLogin == other.userLogin

    override fun hashCode(): Int {
        return this.secret.hashCode()
    }

    companion object {

        private val RANDOM = SecureRandom()

        private fun nextToken(): String {
            return BigInteger(132, RANDOM).toString(64)
        }
    }

    override fun getUserLogin() = this.login

    override fun getCreationDate() = this.creation

    override fun getExpirationDate() = Date(Long.MAX_VALUE)
}
